packages/beacon-node/src/chain/bls/multithread/index.ts

```diff
@@ -346,44 +373,111 @@ export class BlsMultiThreadWorkerPool implements IBlsVerifier {
 
       // Send work package to the worker
       // If the job, metrics or any code below throws: the job will reject never going stale.
       // Only downside is the the job promise may be resolved twice, but that's not an issue
 
       const jobStartNs = process.hrtime.bigint();
-      const workResult = await workerApi.verifyManySignatureSets(jobs.map((job) => job.workReq));
+      const workResult = await workerApi.verifyManySignatureSets(workReqs);
       const jobEndNs = process.hrtime.bigint();
       const {workerId, batchRetries, batchSigsSuccess, workerStartNs, workerEndNs, results} = workResult;
 
       let successCount = 0;
       let errorCount = 0;
 
       // Un-wrap work package
-      for (let i = 0; i < jobs.length; i++) {
-        const job = jobs[i];
+      for (let i = 0; i < jobsStarted.length; i++) {
+        const job = jobsStarted[i];
         const jobResult = results[i];
-        const sigSetCount = job.workReq.sets.length;
-        if (!jobResult) {
-          job.reject(Error(`No jobResult for index ${i}`));
-          errorCount += sigSetCount;
-        } else if (jobResult.code === WorkResultCode.success) {
-          job.resolve(jobResult.result);
-          successCount += sigSetCount;
-        } else {
-          const workerError = Error(jobResult.error.message);
-          if (jobResult.error.stack) workerError.stack = jobResult.error.stack;
-          job.reject(workerError);
-          errorCount += sigSetCount;
+        const sigSetCount = jobItemSigSets(job);
+
+        if (job.type === JobQueueItemType.default) {
+          if (!jobResult || jobResult.code !== WorkResultCode.success) {
+            job.reject(getJobResultError(jobResult, i));
+            errorCount += sigSetCount;
+          } else {
+            job.resolve(jobResult.result);
+            successCount += sigSetCount;
+          }
+        }
+        // handle result of the verification of aggregated signature against aggregated pubkeys
+        else if (job.type === JobQueueItemType.sameMessage) {
+          if (!jobResult || jobResult.code !== WorkResultCode.success) {
+            job.reject(getJobResultError(jobResult, i));
+            errorCount += 1;
+          } else {
+            if (jobResult.result) {
+              // All are valid, most of the time it goes here
+              job.resolve(job.sets.map(() => true));
+            } else {
+              // Retry each individually
+              // Create new jobs for each pubkey set, and Promise.all all the results
+              for (const j of jobItemSameMessageToMultiSet(job)) {
+                if (j.opts.priority) {
+                  this.jobs.unshift(j);
+                } else {
+                  this.jobs.push(j);
+                }
+              }
+              this.metrics?.blsThreadPool.sameMessageRetryJobs.inc(1);
+              this.metrics?.blsThreadPool.sameMessageRetrySets.inc(job.sets.length);
```

Comment:
Is it worth pulling this into a method since this block is repeated? eg: `retryJobItemSameMessage`

Reply:
👍 I extracted it to new method `retryJobItemSameMessage`